Feature/459 more label colors

lynparbone opened the pull request:
Allow the user to configure custom colors by providing a hexadecimal value.

Fixes #459

lynparbone committed:
#459: More label colors

Added input field and button for color input

lynparbone committed:
#459: More label colors

Add Regex validation for the color hex code so no invalid codes can be
set

lynparbone committed:
#459: More label colors

Get contrasting foreground color based on background color

lynparbone committed:
#459: More label colors

Move color validation to component

lynparbone committed:
#459: More label colors

Add TODO to move color functions to own class/ file

lynparbone committed:
#459: More label colors

Update e2e tests to include setting a label color with a hexadecimal
value, cleanup component

lynparbone committed:
#459: More label colors

Add some styling for the color input

lynparbone committed:
#459: More label colors

Refactoring

lynparbone committed:
Merge branch 'develop' into feature/459-more-label-colors

danielsuter commented on scenarioo-client/app/manage/label-colors/color-helpers.ts:
```diff
@@ -0,0 +1,16 @@
+import * as contrast from 'contrast/index';
+
+export const hexadecimalColorIsValid = (color: string): boolean => {
```
convention is more like isBlablaValid

lynparbone replied:
Changed to isHexColorValid, thanks!

danielsuter commented on scenarioo-client/app/manage/label-colors/color-helpers.ts:
```diff
@@ -0,0 +1,16 @@
+    return hexadecimalColorRegexPattern.test(color);
+};
+
+export const getContrastingForegroundColorOfBackgroundColor = (backgroundColor: string): string => {
```
does not matter that it's a background color for this function. I would go with getContrastingColor(hexColor: string)

lynparbone replied:
Done

danielsuter commented on scenarioo-client/test/e2e/pages/labelConfigurationsPage.ts:
```diff
@@ -43,6 +43,19 @@ class LabelConfigurationsPage {
         return this.saveButton.click();
     }
 
+    async updateLabelConfigurationWithColorAsHexadecimal(rowIndex, labelName, colorAsHexadecimal) {
```
add types

lynparbone replied:
Done

lynparbone committed:
#459: More label colors

Rename functions

lynparbone committed:
#459: More label colors

Change layout of color configuration, add label preview

lynparbone committed:
#459: More label colors

Remove contrast package, write own implementation

lynparbone committed:
#459: More label colors

Validate color during input, cleanup

lynparbone committed:
#459: More label colors

Remove unnecessary function, add tooltip for color button

lynparbone committed:
#459: More label colors

Update contrast calculation of colors, general cleanup

lynparbone committed:
#459: More label colors

Update e2e tests

lynparbone committed:
#459: More label colors

Rename functions and variables for color helpers

lynparbone committed:
#459: More label colors

Restructure label colors css

lynparbone committed:
#459: More label colors

Clean up design

lynparbone committed:
#459: More label colors

Cleanup tests and component

lynparbone committed:
Merge branch 'develop' into feature/459-more-label-colors

lynparbone committed:
#459: More label colors

Cleanup, adjust styling, change validation of color

lynparbone committed:
#459: More label colors

Adjust styling so color error is displayed below input

lynparbone commented:
I noticed there is some duplicate code in multiple components regarding the labels.
I've created a separate issue for this here: https://github.com/scenarioo/scenarioo/issues/1016

bruderol commented:
@lynparbone thanks! Perfect to open issues for things that have to be improved later!

bruderol commented on scenarioo-client/test/e2e/specs/configure_label_colors.ts:
```diff
@@ -15,16 +15,15 @@ useCase('Configure label colors')
-
                 // Given number of preconfigured colors (in config.xml for demo)
-                const numberOfPreconfiguredColors = 3;
+                const numberOfPreconfiguredColors: number = 3;
```
type not needed here  - typescript recognises the type when assigning a value

lynparbone replied:
Thanks for the info, will keep that in mind

bruderol commented on scenarioo-client/app/manage/label-colors/color-helpers.ts:
```diff
@@ -0,0 +1,54 @@
+            return extendedHexColor.concat([currentHexValue, currentHexValue]);
+        }, [])
+        .join('');
+};
```
Würde daraus eine Klasse mit statischen Methoden machen und sie Color nennen - Color.isValid(...), Color.getRandom(), und Color.getContrasting(..) sieht hübsch aus und lässt sich einfacher importieren als wenn ich die genauen Methodennamen erst raussuchen und einzeln importieren muss.

bruderol replied:
diese Utility klasse würde ich in ein Utility package legen, falls wir noch mehr mit Farben machen

lynparbone replied:
This makes sense and does simplify the import. Thank you for pointing it out.
I've created a Color class with static functions and placed it in the shared/utils folder.

bruderol commented on scenarioo-client/test/e2e/specs/configure_label_colors.ts:
```diff
@@ -46,4 +45,25 @@ useCase('Configure label colors')
+        scenario('Create a label configuration with a custom color')
+            .it(async () => {
+                const customLabelColor: string = '#00FF00';
+                const numberOfPreconfiguredColors: number = 3;
```
auch hier sind die Types unnötig - Type inference

lynparbone replied:
Thanks for the info.

bruderol commented on scenarioo-client/test/e2e/specs/configure_label_colors.ts:
```diff
@@ -46,4 +45,25 @@ useCase('Configure label colors')
+                await labelConfigurationsPage.assertConfigurationColor(3, customLabelColor);
+                await step('go back to label config page, label is still there and has the custom color');
+
+            });
```
random color button wurde nicht geprüft

lynparbone replied:
Added a check for the random button in the e2e tests.

bruderol reviewed (commented):
Thanks! Great work! See my minor comments for how to slightly improve. Then it is perfect for me.

Maybe avoid that the last empty line for adding a new label can not be removed ??

lynparbone committed:
#459: More label colors

Create utils class Color to replace the color-helper

lynparbone committed:
#459: More label colors

Update tests to include the random color button

lynparbone committed:
#459: More label colors

Add condition for delete button - Delete should only work if the label
configuration isn't empty

lynparbone committed:
#459: More label colors

Use Color utils class in the component and label configurations instead
of the color helper

lynparbone committed:
#459: More label colors

Change tests to check Regex pattern for random color

lynparbone committed:
#459: More label colors

Display delete button only if label name is set

lynparbone committed:
#459: More label colors

Rename function in Colors class

lynparbone committed:
#459: More label colors

Display delete button only if not last line..

lynparbone commented:
Made the delete button conditional: It is now never displayed on the last line, so it can't be deleted by accident (Note: As soon as the user starts typing a label name, it is displayed on that line as a new line is added)

@bruderol: IMO it is ok to merge now

bruderol commented:
Thanks! :+1:
